File: algo/src/test/java/com/hartwig/actin/algo/TestStandardOfCareApplication.java
package com.hartwig.actin.algo;

import java.io.File;
import java.io.IOException;
import java.time.LocalDate;
import java.util.List;
import java.util.Set;
import java.util.stream.Collectors;

import com.hartwig.actin.PatientRecord;
import com.hartwig.actin.PatientRecordFactory;
import com.hartwig.actin.algo.calendar.ReferenceDateProviderTestFactory;
import com.hartwig.actin.algo.doid.DoidConstants;
import com.hartwig.actin.algo.soc.EvaluatedTreatmentInterpreter;
import com.hartwig.actin.algo.soc.RecommendationEngine;
import com.hartwig.actin.algo.soc.TreatmentDB;
import com.hartwig.actin.clinical.datamodel.ClinicalRecord;
import com.hartwig.actin.clinical.datamodel.ImmutableClinicalRecord;
import com.hartwig.actin.clinical.datamodel.ImmutablePriorTumorTreatment;
import com.hartwig.actin.clinical.datamodel.ImmutableTumorDetails;
import com.hartwig.actin.clinical.datamodel.TestClinicalFactory;
import com.hartwig.actin.clinical.datamodel.TreatmentCategory;
import com.hartwig.actin.clinical.datamodel.TumorDetails;
import com.hartwig.actin.clinical.util.ClinicalPrinter;
import com.hartwig.actin.doid.DoidModel;
import com.hartwig.actin.doid.DoidModelFactory;
import com.hartwig.actin.doid.datamodel.DoidEntry;
import com.hartwig.actin.doid.serialization.DoidJson;
import com.hartwig.actin.molecular.datamodel.MolecularRecord;
import com.hartwig.actin.molecular.datamodel.TestMolecularFactory;
import com.hartwig.actin.molecular.util.MolecularPrinter;

import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.jetbrains.annotations.NotNull;

public class TestStandardOfCareApplication {

    private static final Logger LOGGER = LogManager.getLogger(TestStandardOfCareApplication.class);
    private static final String DOID_JSON_PATH =
            String.join(File.separator, System.getProperty("user.home"), "hmf", "repos", "common-resources-public", "doid.json");

    public static void main(@NotNull String... args) throws IOException {
        new TestStandardOfCareApplication().run();
    }

    public void run() throws IOException {
        PatientRecord patient = patient();
        LOGGER.info("Running ACTIN Test SOC Application with clinical record");
        ClinicalPrinter.printRecord(patient.clinical());
        LOGGER.info("and molecular record");
        MolecularPrinter.printRecord(patient.molecular());

        LOGGER.info("Loading DOID tree from {}", DOID_JSON_PATH);
        DoidEntry doidEntry = DoidJson.readDoidOwlEntry(DOID_JSON_PATH);
        LOGGER.info(" Loaded {} nodes", doidEntry.nodes().size());
        DoidModel doidModel = DoidModelFactory.createFromDoidEntry(doidEntry);

        RecommendationEngine recommendationEngine =
                new RecommendationEngine(doidModel, ReferenceDateProviderTestFactory.createCurrentDateProvider());
        EvaluatedTreatmentInterpreter recommendationInterpreter =
                recommendationEngine.provideRecommendations(patient, TreatmentDB.loadTreatments());

        LOGGER.info(recommendationInterpreter.summarize());
        LOGGER.info(recommendationInterpreter.csv());
    }

    private static PatientRecord patient() {
        TumorDetails tumorDetails = ImmutableTumorDetails.builder().addDoids(DoidConstants.COLORECTAL_CANCER_DOID).build();
        ClinicalRecord clinicalRecord = ImmutableClinicalRecord.builder()
                .from(TestClinicalFactory.createMinimalTestClinicalRecord())
                .tumor(tumorDetails)
                .priorTumorTreatments(priorTreatmentStreamFromNames(List.of("CAPOX"), TreatmentCategory.CHEMOTHERAPY))
                .build();
        MolecularRecord molecularRecord = TestMolecularFactory.createProperTestMolecularRecord();

        return PatientRecordFactory.fromInputs(clinicalRecord, molecularRecord);
    }

    private static Set<ImmutablePriorTumorTreatment> priorTreatmentStreamFromNames(List<String> names, TreatmentCategory category) {
        return names.stream()
                .map(treatmentName -> ImmutablePriorTumorTreatment.builder()
                        .name(treatmentName)
                        .isSystemic(true)
                        .startYear(LocalDate.now().getYear())
                        .addCategories(category)
                        .build())
                .collect(Collectors.toSet());
    }
}

File: algo/src/test/kotlin/com/hartwig/actin/algo/soc/TestStandardOfCareApplication.kt
package com.hartwig.actin.algo.soc

import com.hartwig.actin.PatientPrinter
import com.hartwig.actin.algo.doid.DoidConstants
import com.hartwig.actin.algo.evaluation.RuleMappingResourcesTestFactory
import com.hartwig.actin.datamodel.PatientRecord
import com.hartwig.actin.datamodel.TestPatientFactory
import com.hartwig.actin.datamodel.clinical.TreatmentTestFactory
import com.hartwig.actin.doid.DoidModel
import com.hartwig.actin.doid.DoidModelFactory
import com.hartwig.actin.doid.datamodel.DoidEntry
import com.hartwig.actin.doid.serialization.DoidJson
import com.hartwig.actin.icd.IcdModel
import com.hartwig.actin.icd.serialization.CsvReader
import com.hartwig.actin.icd.serialization.IcdDeserializer
import com.hartwig.actin.medication.AtcTree
import com.hartwig.actin.treatment.database.TreatmentDatabase
import com.hartwig.actin.treatment.database.TreatmentDatabaseFactory
import java.io.File
import java.time.LocalDate
import kotlin.system.exitProcess
import io.github.oshai.kotlinlogging.KotlinLogging

class TestStandardOfCareApplication {

    fun run(): Int {
        logger.info { "Running ACTIN Test SOC Application" }

        logger.info { "Loading DOID tree from $DOID_JSON_PATH" }
        val doidEntry: DoidEntry = DoidJson.readDoidOwlEntry(DOID_JSON_PATH)
        logger.info { " Loaded ${doidEntry.nodes.size} nodes" }
        val doidModel: DoidModel = DoidModelFactory.createFromDoidEntry(doidEntry)

        logger.info { "Creating ICD-11 tree from file $ICD_TSV_PATH" }
        val icdNodes = IcdDeserializer.deserialize(CsvReader.readFromFile(ICD_TSV_PATH))
        logger.info { " Loaded ${icdNodes.size} nodes" }
        val icdModel = IcdModel.create(icdNodes)

        val treatmentDatabase = TreatmentDatabaseFactory.createFromPath(TREATMENT_JSON_PATH)

        val standardOfCareEvaluator = StandardOfCareEvaluatorFactory(
            RuleMappingResourcesTestFactory.create(
                doidModel = doidModel,
                icdModel = icdModel,
                atcTree = AtcTree.createFromFile(ATC_TREE_PATH),
                treatmentDatabase = treatmentDatabase
            )
        ).create()

        val patient = patient(treatmentDatabase)
        logger.info { "Generating recommendations for patient record" }
        PatientPrinter.printRecord(patient)

        logger.info { standardOfCareEvaluator.summarizeAvailableTreatments(patient) }
        val patientHasExhaustedStandardOfCare = standardOfCareEvaluator.patientHasExhaustedStandardOfCare(patient)
        logger.info { "Standard of care has${if (patientHasExhaustedStandardOfCare) "" else " not"} been exhausted" }
        return 0
    }

    companion object {
        private val logger = KotlinLogging.logger {}

        private val DOID_JSON_PATH = listOf(
            System.getProperty("user.home"),
            "hmf",
            "repos",
            "actin-resources-private",
            "disease_ontology",
            "doid.json"
        ).joinToString(File.separator)

        private val ACTIN_RESOURCE_PATH = listOf(
            System.getProperty("user.home"),
            "hmf",
            "repos",
            "actin-resources-private"
        ).joinToString(File.separator)

        private val TREATMENT_JSON_PATH = ACTIN_RESOURCE_PATH + File.separator + "treatment_db"

        private val ICD_TSV_PATH = listOf(
            ACTIN_RESOURCE_PATH, "icd", "SimpleTabulation-ICD-11-MMS-en.tsv"
        ).joinToString(File.separator)

        private val ATC_TREE_PATH = listOf(ACTIN_RESOURCE_PATH, "atc_config", "atc_tree.tsv").joinToString(File.separator)

        private fun patient(treatmentDatabase: TreatmentDatabase): PatientRecord {
            val base = TestPatientFactory.createMinimalTestWGSPatientRecord()
            val recentDate = LocalDate.now().minusDays(7)
            return base.copy(
                tumor = base.tumor.copy(doids = setOf(DoidConstants.COLORECTAL_CANCER_DOID), name = "sigmoid"),
                oncologicalHistory = listOf(CAPOX, CAPECITABINE, BEVACIZUMAB, TRIFLURIDINE_TIPIRACIL, FOLFIRI, PANITUMUMAB)
                    .map {
                        val treatment = treatmentDatabase.findTreatmentByName(it) ?: throw IllegalStateException("Treatment not found: $it")
                        TreatmentTestFactory.treatmentHistoryEntry(
                            setOf(TreatmentTestFactory.drugTreatment(treatment.name, treatment.categories().first())),
                            stopYear = recentDate.year,
                            stopMonth = recentDate.monthValue
                        )
                    }
            )
        }
    }
}

fun main(): Unit = exitProcess(TestStandardOfCareApplication().run())
